### PR TITLE
XCOMMONS-2016: Add a new escape method that only escape the minimal amount of XML special characters

### PR DESCRIPTION
**Pull request by manuelleduc**
https://jira.xwiki.org/browse/XCOMMONS-2016

### CODE REVIEW

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 1]
```diff
@@ -274,6 +274,39 @@ public static String escapeAttributeValue(Object content)
...
 
+    /**
+     *  Escapes some XML special characters (&amp; and &lt;) in a <code>String</code> using numerical XML
+     *  entities, so that the resulting string can safely be used in HTML pages.
+     *
+     * @param content the text to escape, may be {@code null}
+     * @return a new escaped {@code String}, {@code null} if {@code null} input
+     * @since 12.8RC1
+     */
+    public static String minimalEscape(Object content)
```
Comment (tmortagne) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:285]: Missing Unstable attribute.

**Review by tmortagne — APPROVED** [review]

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 1 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:287]: Fixed

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 2]
```diff
@@ -274,6 +275,40 @@ public static String escapeAttributeValue(Object content)
...
+     * @since 12.8RC1
+     */
+    @Unstable
+    public static String minimalEscape(Object content)
+    {
+        if (content == null) {
+            return null;
+        }
+        String str = String.valueOf(content);
+        StringBuilder result = new StringBuilder((int) (str.length() * 1.1));
```
Comment (mflorea) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:293]: Any magic number deserves a comment.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 3]
```diff
@@ -274,6 +275,40 @@ public static String escapeAttributeValue(Object content)
...
+        int length = str.length();
```
Comment (mflorea) [inline xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:294]: You're using ``str.length()`` above also so I would declare the variable before.

File: xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/XMLUtilsTest.java [hunk 4]
```diff
@@ -65,6 +65,24 @@ void escape()
...
+    @Test
+    void escapeMinimal()
+    {
+        String escapedText = XMLUtils.minimalEscape("a < a' && a' < a\" => a < a\"");
+
+        assertFalse(escapedText.contains("<"), "Failed to escape <");
+        assertFalse(escapedText.contains("&&"), "Failed to escape &");
+        assertTrue(escapedText.contains(">"), "Failed to keep >");
+        assertTrue(escapedText.contains("'"), "Failed to keep '");
+        assertTrue(escapedText.contains("\""), "Failed to keep \"");
```
Comment (mflorea) [inline xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/XMLUtilsTest.java:77]: I find it safer to assert the exact output and not whether the output contains some strings. It's not like the input string is that long. It's also easier to see the difference between the expected and the actual value when the test fails.

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 2 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: Documented here https://github.com/xwiki/xwiki-commons/pull/101/files#diff-c1eed56adbad8fd67a8e196907ce032fR293

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java (hunk 3 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java]: Factored here https://github.com/xwiki/xwiki-commons/pull/101/files#diff-c1eed56adbad8fd67a8e196907ce032fR295

File: xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/XMLUtilsTest.java (hunk 4 above)
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/XMLUtilsTest.java]: I reused the testing style of the test class, but I agree with you. Test updated here https://github.com/xwiki/xwiki-commons/pull/101/files#diff-4d325ecb5d796d402110ecb9c1edfdf7R72

**Review by mflorea — APPROVED** [review]

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java [hunk 5]
```diff
@@ -274,6 +275,41 @@ public static String escapeAttributeValue(Object content)
...
+                    result.append(AMP);
+                    break;
+                case '<':
+                    result.append(LT);
+                    break;
+                default:
+                    result.append(c);
+            }
+        }
+        return result.toString();
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:310]: I agree this is not the usual way to escape html. But in some context (see https://jira.xwiki.org/browse/XWIKI-9680), if we fully escape the string the content is badly displayed for the end user (IIUC, it is caused by a miss use  of `PropertyClass#displayView` method).
This method is an attempt to find a trade off between the needed string escaping on one side, and the user experience on another side.

[the remark this answers is not in the document]
Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java:310]: Then I'm not sure why we're making this public if it's not the right way it should be...

**Pull request merged**